File: app/src/main/java/com/king/anetty/app/MainActivity.java
/*
 * Copyright 2019 Jenly Yu
 * <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <a href="https://github.com/jenly1314">jenly1314</a>
 *
 * The Netty Project licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
package com.king.anetty.app;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.king.anetty.ANetty;
import com.king.anetty.Netty;

public class MainActivity extends AppCompatActivity {

    private EditText etHost;
    private EditText etPort;
    private EditText etContent;
    private TextView tvContent;

    private Toast mToast;

    //TODO 配置您的TCP连接主机
    private String mHost = "192.168.100.48";
    //TODO 配置您的TCP连接端口
    private int mPort = 6000;

    private Netty mNetty;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        etHost = findViewById(R.id.etHost);
        etPort = findViewById(R.id.etPort);
        etContent = findViewById(R.id.etContent);
        tvContent = findViewById(R.id.tvContent);
        tvContent.setMovementMethod(ScrollingMovementMethod.getInstance());
        etHost.setText(mHost);
        etPort.setText(String.valueOf(mPort));
        initNetty();
    }

    /**
     * 初始化Netty
     */
    private void initNetty(){
        mNetty = new ANetty(new Netty.OnChannelHandler() {
            @Override
            public void onMessageReceived(String msg) {
                //接收到消息
                tvContent.append(msg + "\n");
            }

            @Override
            public void onExceptionCaught(Throwable e) {

            }
        }, true);

        mNetty.setOnConnectListener(new Netty.OnConnectListener() {
            @Override
            public void onSuccess() {
                //TODO 连接成功
                showToast("连接成功");
            }

            @Override
            public void onFailed() {
                //TODO 连接失败
                showToast("连接失败");
            }

            @Override
            public void onError(Exception e) {
                //TODO 连接异常
                showToast("连接异常");
                Log.e(ANetty.TAG,e.getMessage());
            }
        });

    }

    /**
     * 连接Netty
     */
    private void connectNetty(){

        mHost = etHost.getText().toString();
        if(!TextUtils.isEmpty(etPort.getText())){
            mPort = Integer.parseInt(etPort.getText().toString());
        }

        if(TextUtils.isEmpty(mHost)){
            showToast("请配置TCP连接的主机");
            return;
        }
        //连接Netty
        mNetty.connect(mHost,mPort);
    }

    /**
     * 发送消息
     * @param msg
     */
    private void sendNettyMsg(String msg){
        if(mNetty.isConnected()){
            mNetty.sendMessage(msg);
        }else{
            showToast("Netty未连接");
        }

    }

    private void showToast(String text){
        if(mToast == null){
            mToast = Toast.makeText(MainActivity.this,text,Toast.LENGTH_SHORT);
        }
        mToast.setText(text);
        mToast.show();

    }

    @Override
    protected void onDestroy() {
        mNetty.disconnect();
        super.onDestroy();
    }

    private void clickSend(){
        if(TextUtils.isEmpty(etContent.getText())){
            showToast("请输入消息内容");
            return;
        }

        sendNettyMsg(etContent.getText().toString());

//        String json = "{\"body\":{\"cNo\":\"粤BD82700\",\"dPermit\":\"132627\",\"lat\":22.53421,\"lon\":114.09242,\"mac\":\"15e1f216f0053c6999968d00480a27b17d2a1a33f0e907fcfd54b360eed1e474\",\"time\":1553562614264},\"head\":{\"sn\":1,\"time\":1553562614265,\"type\":\"LOGIN_REQ\",\"ver\":\"1.0.1\"}}";
//        sendNettyMsg(json);
    }

    public void onClick(View v){
        switch (v.getId()){
            case R.id.btnConnect:
                connectNetty();
                break;
            case R.id.btnSend:
                clickSend();
                break;
        }
    }
}

File: app/src/main/java/com/king/anetty/app/MainActivity.kt
package com.king.anetty.app

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import com.king.anetty.ANetty
import com.king.anetty.Netty.OnChannelHandler
import com.king.anetty.Netty.OnConnectListener
import com.king.anetty.Netty.OnSendMessageListener
import com.king.anetty.app.databinding.ActivityMainBinding
import io.netty.channel.ChannelHandlerContext

/**
 * ANetty 使用示例
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
class MainActivity : AppCompatActivity() {

    private val binding by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    var mHost = "192.168.100.100"

    var mPort = 6000

    lateinit var aNetty: ANetty

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        binding.etHost.setText(mHost)
        binding.etPort.setText(mPort.toString())

        initANetty()
    }

    override fun onDestroy() {
        aNetty.close()
        super.onDestroy()
    }

    private fun getContext() = this

    /**
     * 初始化ANetty
     */
    private fun initANetty() {
        aNetty = ANetty(object : OnChannelHandler {
            override fun onMessageReceived(ctx: ChannelHandlerContext?, msg: String?) {
                binding.tvContent.append("接收：$msg\n")
            }

            override fun onExceptionCaught(ctx: ChannelHandlerContext?, e: Throwable?) {
                binding.tvContent.append("断开连接：$e\n")
                if (!aNetty.isConnected) {
                    binding.btnStart.text = "连接"
                    binding.btnStart.isEnabled = true
                }
            }

        }, true)

        aNetty.setOnConnectListener(object : OnConnectListener {
            override fun onSuccess() {
                binding.tvContent.append("连接成功\n")
                binding.btnStart.text = "断开"
                binding.btnStart.isEnabled = true
                binding.progressBar.isVisible = false
                Toast.makeText(getContext(), "连接成功", Toast.LENGTH_SHORT).show()
            }

            override fun onFailure(e: Throwable?) {
                e?.message?.also {
                    binding.tvContent.append("连接失败：$it\n")
                }
                binding.btnStart.text = "连接"
                binding.btnStart.isEnabled = true
                binding.progressBar.isVisible = false
                Toast.makeText(getContext(), "连接失败", Toast.LENGTH_SHORT).show()
            }
        })

        aNetty.setOnSendMessageListener(object : OnSendMessageListener {
            override fun onSendMessage(msg: Any?) {
                binding.tvContent.append("发送：$msg\n")
            }

            override fun onException(e: Throwable?) {
                if (!aNetty.isConnected) {
                    binding.tvContent.append("断开连接：$e\n")
                    binding.btnStart.text = "连接"
                    binding.btnStart.isEnabled = true
                }
            }
        })
    }

    /**
     * 连接或断开连接
     */
    private fun clickStart() {
        if (!aNetty.isConnected) {
            mHost = binding.etHost.text.toString()
            if (TextUtils.isEmpty(binding.etHost.text)) {
                Toast.makeText(getContext(), "请输入IP地址", Toast.LENGTH_SHORT).show()
                return
            }
            if (TextUtils.isEmpty(binding.etPort.text)) {
                Toast.makeText(getContext(), "请输入端口", Toast.LENGTH_SHORT).show()
                return
            }
            mPort = binding.etPort.text.toString().toInt()
            binding.btnStart.text = "连接中…"
            binding.btnStart.isEnabled = false
            binding.progressBar.isVisible = true
            aNetty.connect(mHost, mPort)
        } else {
            aNetty.disconnectBlocking()
            binding.tvContent.append("断开连接\n")
            binding.btnStart.text = "连接"
            binding.btnStart.isEnabled = true
        }
    }

    /**
     * 发送
     */
    private fun clickSend() {
        if (!TextUtils.isEmpty(binding.etContent.text)) {
            if (aNetty.isConnected) {
                val data = binding.etContent.text.toString()
                aNetty.sendMessage(data)
                binding.etContent.setText("")
            } else {
                Toast.makeText(getContext(), "未连接", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun clickClear() {
        binding.tvContent.text = ""
    }

    fun onClick(v: View) {
        when (v.id) {
            R.id.btnStart -> clickStart()
            R.id.btnSend -> clickSend()
            R.id.btnClear -> clickClear()
        }
    }
}